src/Time/History.hpp

```diff
@@ -3,400 +3,910 @@
+  };
+
+  class MutableSubsteps
+      : public stl_boilerplate::RandomAccessSequence<MutableSubsteps,
+                                                     StepRecord<Vars>> {
+   public:
+    MutableSubsteps() = delete;
+    MutableSubsteps(const MutableSubsteps&) = delete;
+    MutableSubsteps(MutableSubsteps&&) = default;
+    MutableSubsteps& operator=(const MutableSubsteps&) = delete;
+    MutableSubsteps& operator=(MutableSubsteps&&) = default;
+    ~MutableSubsteps() = default;
+
+    size_t size() const { return history_->substep_values_.size(); }
+    static constexpr size_t max_size() { return history_max_substeps; }
+
+    StepRecord<Vars>& operator[](const size_t index) const {
+      ASSERT(index < size(),
+             "Requested substep " << index << " but only have " << size());
+      return history_->substep_values_[index];
+    }
+
+   private:
+    friend History;
+    explicit MutableSubsteps(const gsl::not_null<History*> history)
+        : history_(history) {}
+
+    gsl::not_null<History*> history_;
+  };
+  /// \endcond
+
+  /// Immutable, type-erased access to the history.  This method
+  /// returns a class derived from `ConstUntypedHistory<UntypedVars>`.
+  /// Any modifications to this class invalidate the returned object.
+  ConstUntypedAccess untyped() const { return ConstUntypedAccess(*this); }
+
+  /// Mutable, type-erased access to the history.  This method returns
+  /// a class derived from `MutableUntypedHistory<UntypedVars>`.  Any
+  /// modifications to this class invalidate the returned object,
+  /// except for those performed through the object itself.
+  MutableUntypedAccess untyped() { return MutableUntypedAccess(this); }
+
+  /// Get or set the order the time stepper is running at.  Many time
+  /// steppers expect this to have a particular value.  This has no
+  /// effect on the storage of past data.
+  /// @{
+  size_t integration_order() const { return integration_order_; }
+  void integration_order(const size_t new_integration_order) {
+    integration_order_ = new_integration_order;
+  }
+  /// @}
 
-  HistoryIterator(const UntypedHistory<T>* const history,
-                  const difference_type offset);
+  /// Type and value used to indicate that a record is to be created
+  /// without the `value` field set.
+  /// @{
+  struct NoValue {};
+  static constexpr NoValue no_value{};
+  /// @}
 
-  const UntypedHistory<T>* history_;
-  difference_type offset_;
-};
+  /// Insert a new entry into the history.  It will be inserted as a
+  /// step or substep as appropriate.
+  ///
+  /// The supplied `time_step_id` must be newer than the current
+  /// newest entry, and if the substep of `time_step_id` is nonzero,
+  /// the id must be consistent with the existing data for the current
+  /// step.
+  ///
+  /// If `no_value` is passed, the created record will not have its
+  /// `value` field set.  This is primarily useful for histories for
+  /// the implicit portions of IMEX time steppers.
```

Comment:
This is all in a bit of flux at the moment, unfortunately.  As of this PR, the value field is never used and all the processing of it is actually wasted effort (but much of it is still being done).  The next PR in this sequence will change that and setting it will always be required.

The intended use case is for IMEX, but unfortunately the details are along the lines of "when calling some implicit-integration-related functions that don't exist on develop yet and haven't had their names or interfaces fully nailed down".  (I almost left these functions out of this PR.)  I can add something about stiff source terms if you'd like, but I've now encountered two situations where the solve isn't iterative, so I want to avoid that part.  (These are (1) getting an approximate solution from a semi-implicit solve, which is roughly a normal solve but stopping after one iteration, and (2) the case where the implicit problem can actually be solved analytically, which happens in some regimes for FFE.)

Implementation-wise, integration is split up like `y1 = y0 + dt * explicit_stuff + dt * other_stuff` (where `other_stuff` is just the implicit solve, at least for now).  As of a not-yet-submitted PR, the `value` field is used to obtain `y0`, and the normal TimeStepper interface *sets* `y1` to the first two terms together.  The implicit solve then *adds* its adjustment to the result.  Obviously, adding another constant term would be bad (i.e., zeroth order), so nothing except the main integration needs the `value` field.

After thinking through all that, maybe a good explanation is something about histories that do not represent the primary integration, but rather some adjustment on top of it.  I've made an attempt at adding something.

Reply:
I like what you've put :) I think that helps a lot!